mmengine/model/wrappers/fully_sharded_distributed.py

```diff
@@ -219,5 +253,33 @@ def test_step(self, data: dict) -> List[BaseDataElement]:
         Returns:
             List[BaseDataElement]: The predictions of given data.
         """
-        inputs, data_sample = self.module.data_preprocessor(data, False)
-        return self(inputs, data_sample, mode='predict')
+        data = self.module.data_preprocessor(data, False)
+        return self._run_forward(data, mode='predict')  # type: ignore
+
+    def _run_forward(self, data: Union[dict, tuple, list],
+                     mode: str) -> Union[Dict[str, torch.Tensor], list]:
+        """Unpacks data for :meth:`forward`
+
+        Args:
+            data (dict or tuple or list): Data sampled from dataset.
+            mode (str): Mode of forward.
+
+        Returns:
+            dict or list: Results of training or testing mode.
+        """
+        if isinstance(data, dict):
+            results = self(**data, mode=mode)
+        elif isinstance(data, (list, tuple)):
+            results = self(*data, mode=mode)
+        else:
+            raise TypeError('Output of `data_preprocessor` should be '
+                            f'list, tuple or dict, but got {type(data)}')
+        return results
+
+    def _collect_state_dict(self):
+        full_state_dict_config = FullStateDictConfig(
+            offload_to_cpu=True, rank0_only=True)
+        with FSDP.state_dict_type(self, StateDictType.FULL_STATE_DICT,
```

Comment:
Collecting the whole state dict of a large model every time during saving will be slow. Can we reimplement a unique CheckpointHook for FSDP and only collect the parameters after the training ends?

Reply:
Is there any reference solution? What is the official best practice provided by PyTorch?

Reply:
PyTorch provides 3 methods for saving FSDP checkpoints, as shown [here](https://github.com/pytorch/pytorch/blob/664058fa83f1d8eede5d66418abff6e20bd76ca8/torch/distributed/fsdp/fully_sharded_data_parallel.py#L1893). In this PR I just implemented the most usual & compatible way of saving checkpoint. If we save local/sharded checkpoints, the `load` method should also be changed.

I plan to support other ways of saving FSDP checkpoints in future PRs, together with checkpoint loading/resume. The suggestion of implementing a unique CheckpointHook makes sense to me and I will do it this way!